### PR TITLE
[0.20] [Path] Refactor `Extensions` code out of PocketShape [2 of 3 per dev conversation]

### PR DESCRIPTION
**Pull request by Russ4262**
This PR relocates the `Extensions` code base from PocketShape into independent modules in a `features` subfolder within PathScripts - this is a partial replacement for PR #4324 .  This will make the code easier to access for other operations, say the Adaptive operation.  This PR does not alter the original `Extensions` codebase except for updating method/function calls due to relocation of those methods/functions being called.  Otherwise, the code is intact as prior to this PR.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

File: src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
+# -*- coding: utf-8 -*-
+# ***************************************************************************
+# *   Copyright (c) 2019 Markus Lampert (mlamptert)                         *
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py:3]: please change to sliptonic

File: src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+
+from PySide import QtCore, QtGui
+from pivy import coin
+
+# lazily loaded modules
+from lazy_loader.lazy_loader import LazyLoader
+Part = LazyLoader('Part', globals(), 'Part')
+
+__title__ = "Path Feature Extensions UI"
+__author__ = "Markus Lampert (mlampert)"
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py:39]: same, sliptonic

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    if not hasattr(obj, 'ExtensionFeature'):
+        obj.addProperty('App::PropertyLinkSubListGlobal', 'ExtensionFeature', 'Extension', QtCore.QT_TRANSLATE_NOOP('PathPocketShape', 'List of features to extend.'))
+    if not hasattr(obj, 'ExtensionCorners'):
+        obj.addProperty('App::PropertyBool', 'ExtensionCorners', 'Extension', QtCore.QT_TRANSLATE_NOOP('PathPocketShape', 'When enabled connected extension edges are combined to wires.'))
+        obj.ExtensionCorners = True
+
+    obj.setEditorMode('ExtensionFeature', 2)
+
+
+def set_default_property_values(obj, job):
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/features/PathFeatureExtensions.py:295]: Everything in Path is camel case, please let's stick to that.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+                    return None
+
+            #    return self._extendEdge(feature, edge, direction)
+            return self._extendEdge(feature, edges[0], direction)
+
+        return extendWire(feature, sub, self.length.Value)
+# Eclass
+
+
+def initialize_properties(obj):
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/features/PathFeatureExtensions.py:282]: Please don't use underscore naming convention, everything else is in camel case so let's stick with that.

**Review by mlampert — CHANGES_REQUESTED** [review]

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py (hunk 3 above)
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py:295]: My apologies.  I was attempting to adhere to conventions expressed at [Path Code Style Guide](https://forum.freecadweb.org/viewtopic.php?f=15&t=46845).  I will change them.

Thank you for the review.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
+# -*- coding: utf-8 -*-
+# ***************************************************************************
+# *   Copyright (c) 2019 Markus Lampert (mlamptert)                         *
```
[the remark this answers is not in the document]
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py:3]: Okay, to this instance and the others likewise requested.

[the remark this answers is not in the document]
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py]: Done.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+import math
+
+# lazily loaded modules
+from lazy_loader.lazy_loader import LazyLoader
+Part = LazyLoader('Part', globals(), 'Part')
+
+from PySide import QtCore
+
+__title__ = "Path Features Extensions"
+__author__ = "Markus Lampert (mlampert)"
```
[the remark this answers is not in the document]
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py]: Done

File: src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py (hunk 1 above)
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py]: Done.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py (hunk 2 above)
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensionsGui.py]: Done.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py (hunk 3 above)
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py]: Done.

File: src/Mod/Path/PathScripts/features/PathFeatureExtensions.py (hunk 4 above)
Comment (Russ4262) [reply src/Mod/Path/PathScripts/features/PathFeatureExtensions.py]: Done.

**Comment by Russ4262** [comment]
This PR will require modifications after PR #4617 is merged, as the code in that PR will have to be added to this PR due to relocation of the code.

**Comment by Russ4262** [comment]
Closing in light of recent discussion with @sliptonic regarding PR #4388 .  Thanks MLampert for your review and support.

**Pull request closed without merging**